Git history of project (#63)
arcus/education_modules: git_history_of_project/git_history_of_project.md

<!--

author:  Elizabeth Drellich
email:    drelliche@chop.edu

> version:  1.0.0

Reviewer: ```suggestion
version:  1.0.0
module_template_version: 2.0.0
```

long_description: You know that version control is important. You know how to save your work to your Git repository. Now you are ready to look at and compare different versions of your work. In this module you will you will learn how to navigate through the commits you have made to Git. You will also learn how to compare current code with past code, and, if necessary, revert to an earlier version of your work.

@estimated_time: 50 minutes

@learning_objectives  
After completion of this module, learners will be able to:

- Identify and use the HEAD of a repository.
- Identify and use Git commit numbers.
- Compare versions of tracked files.
- Restore old versions of files.

@end

link:  https://chop-dbhi-arcus-education-website-assets.s3.amazonaws.com/css/styles.css

script: https://kit.fontawesome.com/83b2343bd4.js

-->

# Exploring the History of your Git Repository

<div class = "overview">

## Overview
@comment

**Is this module right for me?** @long_description

**Estimated time to completion:**
@estimated_time

**Pre-requisites**

This module is for you if:

> * You know know how to create a Git repository and make changes using `add` and `commit` from a command line interface (CLI)

Reviewer: ...Which you can learn in `[git create](link)`

* You want to learn how to compare different versions of your work and restore files to previous versions.

**Learning Objectives**

@learning_objectives

</div>

## Lesson Preparation

> The purpose of this module is to learn how to explore the history of a Git repository that has already been created. We are going to download Dracula's repository in which he started documenting the pros and cons of moving with his friends to Mars. Open up your command line interface and type:

Reviewer: ```suggestion
The purpose of this module is to learn how to explore the history of a Git repository that has already been created. We are going to download Dracula's repository in which he started documenting the pros and cons of moving with his friends to Mars. Open up your command line interface, navigate to the directory you'd like to clone a repository into, and type:
```

Also at some point around here, you should copy/paste the "but if you're in windows..." caveat from git create/track

```
git clone https://www.github.com/arcus/planets
```

> This should be a very fast download. Now if you type `ls` you should see a folder titled `planets`. Navigate into the planets directory. This is the directory we will be exploring throughout the module.

Reviewer: ```suggestion
This should be a very fast download. Now if you type `ls` within the repository directory, you should see a folder titled `planets`. Navigate into the planets directory using `cd`. This is the directory we will be exploring throughout the module.
```

> <div class = "warning">

Reviewer: Whoops haha I should have waited.  But since I've suggested linux commands ahead of this, maybe move it up?

If you are on a Windows machine, make sure you are in the Git Bash application. The Command Prompt (cmd) will not recognize the commands we are using in this module, including `ls`. 
</div>

![A collage of the project we will use as a motivating example. Figures include Dracula, a werewolf, a mummy, and several planets and moons.](media/motivatingexample.png)
[Werewolf vs dracula](https://www.deviantart.com/b-maze/art/Werewolf-vs-Dracula-124893530)

> by [b-maze](https://www.deviantart.com/b-maze) / [Deviant Art](https://www.deviantart.com/).

Reviewer: same concern about b-maze's portfolio being a wee bit NSFW

[Mars](https://en.wikipedia.org/wiki/File:OSIRIS_Mars_true_color.jpg) by European Space Agency /
[CC-BY-SA 3.0 IGO](https://creativecommons.org/licenses/by/3.0/deed.en).
[Pluto](https://commons.wikimedia.org/wiki/File:PIA19873-Pluto-NewHorizons-FlyingPastImage-20150714-transparent.png) /
Courtesy NASA/JPL-Caltech.
[Mummy](https://commons.wikimedia.org/wiki/File:Mummy_icon_-_Noun_Project_4070.svg)

> $\copyright$; Gilad Fried / [The Noun Project](https://thenounproject.com/) /

Reviewer: ```suggestion
$\copyright$ Gilad Fried / [The Noun Project](https://thenounproject.com/) /
```

[CC BY 3.0](https://creativecommons.org/licenses/by/3.0/deed.en).
[Moon](https://commons.wikimedia.org/wiki/File:Lune_ico.png)

> $\copyright$; Luc Viatour / [https://lucnix.be](https://lucnix.be/) /

Reviewer: ```suggestion
$\copyright$ Luc Viatour / [https://lucnix.be](https://lucnix.be/) /
```

[CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/deed.en).

<div class = "options">

> If you have a repository with at least a few commits in it, such as the one created in the previous module in this series, you can use this module as a framework for exploring that instead.

Reviewer: ```suggestion
If you have a repository with at least a few commits in it, such as the one created in the previous module in this series, you can use this module as a framework for exploring that instead.  Just keep in mind that our instructions include adding new files and new lines, so think about whether you want to work in an existing repository (maybe you know you need to make some edits there anyway) or would rather not tinker with something that's important to you.
```

Alternatively, you can clone any repository you find on [GitHub](https://github.com) using

```
git clone https://www.github.com/RepositoryOwner/RepositoryName
```

> and explore how it was created!

Reviewer: Honestly I'd omit the offer now that I think more about it

### Using HEAD

The `HEAD` refers to the most recently commit to your repository. It won't include any changes that haven't been committed to the repository yet. You can see that most recent commit by entering `git show HEAD`.

```
$ git show HEAD
commit 584977a18319f539f3f5ecde6176db2ba0081cd9 (HEAD -> main, origin/main)
Author: Vlad Dracula <vlad@tran.sylvan.ia>
Date:   Mon Mar 14 14:59:40 2022 -0400

    Ignore data files and the results folder.

diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..f451386
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,2 @@
+*.dat
+results/
```

The last thing that Dracula committed to the planets repository was the `.gitignore` file in which he told Git to "Ignore data files and the results folder."

<div class = "care">
Sometimes a command will output many lines of code that may not be relevant to you. For example the `show` command will output not just the commit message, but also every single change that you made in that commit.

The more you see this type of output, the more comfortable you will get with extracting the information you need and ignoring the rest.
</div>

<div class = "important">

> If the output of a command is longer than the number of lines your console displays, you can navigate that output using the down and up arrows on your keyboard, or press `q` to skip to the end of that output.

Reviewer: doesn't `q` actually just "quit" you out without moving ahead to the end of the output?

</div>

Maybe you want to look one step further back into your work. By using `HEAD~n` you can look back n checkpoints in your repository. The `~` symbol is named "tilde" and pronounced "TIL-duh."

For example, to look back just one commit before the most recent checkpoint, use `git show HEAD~1`. If instead you wanted to look back three checkpoints (two commits before the most recent commit) you would enter `git show HEAD~2`.

![The stack of 3 white flat boxes. Now the top box is labeled `HEAD`, the middle box `HEAD~1` and the bottom `HEAD~2`](media/Commit_stack_HEAD.svg)

Using `HEAD` to refer to your commits can be great for looking at recent versions of your repository. But be careful, each time you commit to the repository, you are creating a new most-recent-version that is now the `HEAD` and every older version moves one layer down in the stack.

> ![On the left, the stack of 3 flat white boxes are labeled `HEAD`, `HEAD~1`,  and `HEAD~2`. An arrow points to a stack on the right made of the 3 white boxes with a 4th red box now on top. This red box is labeled `HEAD` and the boxes below it are labeled, in order from top to bottom, `HEAD~1`, `HEAD~2`, `HEAD~3`](media/Commit_stack_HEAD2.svg)

Reviewer: ```suggestion
![On the left, the stack of 3 flat white boxes are labeled `HEAD`, `HEAD~1`,  and `HEAD~2`. An arrow points to a stack on the right made of the 3 white boxes with a 4th red box now on top. This red box is now the one labeled `HEAD` and the boxes below it are labeled, in order from top to bottom, `HEAD~1`, `HEAD~2`, `HEAD~3`, with all the white boxes having changed labels.](media/Commit_stack_HEAD2.svg)
```

I'm not great at alt texts so feel free to veto the suggestion.  ALSO, time saver (but no need to change it)... you can just put backticks around the entire alt text instead of doing on individual words... the output is the same.

When you enter `git commit -m "short descriptive message"` no output appears, but Git gives that commit a unique identifier, its commit number. By typing `git log` we can see ALL of the previous commits. Since that might be a lot of output to deal with, we will just ask for the last 3 commits using `git log -n 3`. The number after `-n` is how many commits you want displayed from the log.

Let's see what that output looks like for Dracula:

```
$ git log -n 3
commit 584977a18319f539f3f5ecde6176db2ba0081cd9 (HEAD -> main, origin/main)
Author: Vlad Dracula <vlad@tran.sylvan.ia>
Date:   Mon Mar 14 14:59:40 2022 -0400

    Ignore data files and the results folder.

commit fe532b097861acb8bd3d7f221d6ee741249dc8f0
Author: Vlad Dracula <vlad@tran.sylvan.ia>
Date:   Mon Mar 14 14:58:17 2022 -0400

    Add concerns about effects of Mars' moons on Wolfman

commit c3fe500d96fa9114210320f3cac0a0d9fd58d33c
Author: Vlad Dracula <vlad@tran.sylvan.ia>
Date:   Mon Mar 14 14:57:09 2022 -0400

    Start notes on Mars as a base
```

> If you forget to include the `-n 3` flag and just type `git log` you can always jump to the end by pressing `q`. If you have been working on a project for a while, using the arrow keys to scroll down may take a prohibitively long time.

Reviewer: I don't think this is true?  I think q just quits?

> The first thing to notice is that all of your commit messages are here. This is a good reminder to write clear and concise messages because future you may be very grateful when trying to figure out where exactly past you introduced a particular issue.

Reviewer: ```suggestion
The first thing to notice is that all of your commit messages are here. This is a good reminder to write clear and concise messages because **future** you may be very grateful when trying to figure out where exactly **past** you introduced a particular issue.
```

The second thing to notice is the structure of each entry in the log: commit, Author, Date, message.

> When you identify which commit you want to look at, the commit number is the unique 40 digit string of letters and numbers above it after the word "commit". In Dracula's repository, the unique identifier for the commit in which he "add[ed] concerns about effects of Mars' moons on Wolfman" is `fe532b097861acb8bd3d7f221d6ee741249dc8f0`.

Reviewer: "digit" is true, because the letters here are actually base 16 digits.  Another interesting mathy aside that may be too far afield to put here but in case some people are like "hmmm why 'digit', those are letters not numbers"...

<div class = "warning">
If you are using the repo you created in an earlier module,  your commits might have different commit numbers than the commit numbers in the examples. Be sure to use the commit number in YOUR log.
</div>

> Now that we have found the unique identifier of the commit we want to examine, we can use it with `git show`. Don't worry, you won't need to type all 40 digits, the first six will suffice.

Reviewer: ```suggestion
Now that we have found the unique identifier of the commit we want to examine, we can use it with `git show`. Don't worry, you won't need to type all 40 digits, the first six will suffice in almost every case.  It's very, very unlikely that you'll have two commit codes that share the same first six digits.  Because the first six digits are almost always unique, you'll see that shortened version a lot in git.
```
Trying to explain a bit more, let me know if this feels like overkill.

When you commit code to Git, it doesn't create a brand new copy of your files from scratch. Instead it records the differences between the version of each file that it already has stored, and the new version.

You could think of each commit not as a checkpoint, but the instructions on how to get to the next checkpoint. The entire repository can be built by following a series of instructions (commits) in order. Similarly you can determine the precise state of an earlier version by following the instructions backwards.

![The three white boxes are shown with a large green arrow pointing from each box to the box above it. The top box is labeled "Last committed version of repository" and the arrow pointing to it from the middle box is labeled "HEAD." The second box is labeled "Next to last version of repository" and the arrow pointing to it from the box below is labeled "HEAD~1." The bottom box is labeled "Previous version of repository" and the arrow pointing to it is labeled "HEAD~2." There are three dots under the boxes, and another three dots under the arrows indicating that this pattern continues.](media/Commits_are_pointers.svg)
</div>

## Comparing files to prior commits

> In order to have some more commits to compare, let's add a few more lines to Dracula's notes about Mars.

Reviewer: ```suggestion
In order to have some more commits to compare, let's add a few more lines to Dracula's notes about Mars.  If you're working in your own project, and not the "planets" repo, you can still do hands-on work here. 
 Instead of using our suggestions for changes, try making small changes that won't break anything important, like adding an extra line between two paragraphs of text or correcting a typo in a code comment you've been meaning to get to.
```

But the Mummy will appreciate the lack of humidity
```
Let's add and commit this change:

```
$ git add mars.txt
$ git commit -m "Discuss concerns about Mars' climate for Mummy"
```

Next let's add one more line to `mars.txt` so that it reads:

```
Cold and dry, but everything is my favorite color
The two moons may be a problem for Wolfman
But the Mummy will appreciate the lack of humidity
An ill-considered change
```

Notice that we did NOT add or commit that last change to `mars.txt`. When you are working on a large project, it is very easy to lose track in your head of when you last committed, or what changes you have made since. In this section we will see how to ask Git to tell us that information.

<div class = "learnmore">
You can change things in the `planets` directory that you downloaded from GitHub!

Because these files now exist on your computer, you can change them and commit those changes. The author of the new commits will be you, not Vlad Dracula. You can use `git log` to see which of you made which commit.

This is a preview of how Git lets you work with collaborators to create a project together.
</div>

### Using `diff`

> You might have noticed that `diff` was the first word after your commit message when you entered `git show HEAD`. The lines after that are **all** the changes that you made with that commit. When comparing earlier versions with the current version, you are usually only going to want to look at one file at  a time. By typing `git diff HEAD` you are asking to see all the differences between the current state of your tracked files and the most recently committed version. Let's give it a try:

Reviewer: I'm not sure why HEAD is getting so much attention (from Carpentries I mean) because it can move!  Maybe toward the top you note that "checkout" is a HEAD mover, and do a checkout to make sure HEAD is on main?  This will help understand why HEAD is important. https://work.njae.me.uk/2020/06/23/git-for-beginners-part-3-moving-head/

```
$ git diff HEAD
diff --git a/mars.txt b/mars.txt
index b36abfd..93a3e13 100644
--- a/mars.txt
+++ b/mars.txt
@@ -1,3 +1,4 @@
 Cold and dry, but everything is my favorite color
 The two moons may be a problem for Wolfman
 But the Mummy will appreciate the lack of humidity
+An ill-considered change
```

There is a lot of information here so we will go through what each of the parts means.

```
diff --git a/mars.txt b/mars.txt
```
1. The first line tells us that Git is producing output similar to the Unix `diff` command comparing the old and new versions of the file.

> 

Reviewer: I can't remember if you've already used "diff" without explaining it, but at some point it will be useful to be very explicit.

[[ ]] 3. `$ git checkout HEAD~1 data_cruncher.py`
[[X]] 4. `$ git checkout <unique ID of last commit> data_cruncher.py`
[[?]] More than one answer can be correct.
***
<div class = "answer">
The `checkout` command restores files from the repository, overwriting the files in your working directory. Answers 2 and 4 both restore the latest version in the repository of the file `data_cruncher.py`. Answer 2 uses `HEAD` to indicate the latest, whereas answer 4 uses the unique ID of the last commit, which is what `HEAD` means.

Answer 3 gets the version of `data_cruncher.py` from the commit before `HEAD`, which is NOT what we wanted.

Answer 1 will result in a detached `HEAD`! Use `git checkout main` to reattach `HEAD` before you make any more changes.
</div>
***

## Additional Resources

This module was based on examples from [Software Carpentry](https://software-carpentry.org)'s [lessons on using Git](https://swcarpentry.github.io/git-novice/). The story of Dracula and Wolfman's planned trip to Mars is borrowed from them, and more examples are available in the original lesson:

 - Lesson 5: [Exploring History](https://swcarpentry.github.io/git-novice/05-history/index.html)

> This module is also the fourth in a series of modules about Git created by Arcus Education and friends. We suggest continuing with the next module in the series on [working with others in Git](link/goes/here) and the entire collection is available on the [Arcus website](link/goes/here)

Reviewer: Hmmm... Arcus doesn't mean much to most of our eventual audience.  
```suggestion
This module is also the fourth in a series of modules about Git. We suggest continuing with the next module in the series on [working with others in Git](link/goes/here) and the entire collection is available on the [module catalog](link/goes/here)
```
